File: java/src/main/java/helptree/HelpTree.java
package helptree;

import picocli.CommandLine;
import picocli.CommandLine.Model.CommandSpec;
import picocli.CommandLine.Model.OptionSpec;
import picocli.CommandLine.Model.PositionalParamSpec;

import java.util.ArrayList;
import java.util.List;

public class HelpTree {

    public static class Config {
        public boolean helpTree = false;
        public Integer depthLimit = null;
        public List<String> ignoreNames = new ArrayList<>();
        public boolean showAll = false;
        public String outputFormat = "text";
        public String style = "rich";
        public String color = "auto";
        public List<String> path = new ArrayList<>();
        public List<String> remainingArgs = new ArrayList<>();
    }

    public static List<TreeOption> discoveryOptions() {
        return List.of(
            new TreeOption("help-tree", null, "--help-tree", "Enable tree output", false, false, false),
            new TreeOption("tree-depth", "-L", "--tree-depth", "Max recursion depth", false, true, false),
            new TreeOption("tree-ignore", "-I", "--tree-ignore", "Exclude subcommand names", false, true, false),
            new TreeOption("tree-all", "-a", "--tree-all", "Include hidden subcommands/arguments", false, false, false),
            new TreeOption("tree-output", null, "--tree-output", "Output format (text/json)", false, true, false),
            new TreeOption("tree-style", null, "--tree-style", "Output style (plain/rich)", false, true, false),
            new TreeOption("tree-color", null, "--tree-color", "Color control (auto/always/never)", false, true, false)
        );
    }

    public static Config extractConfig(String[] args) {
        Config config = new Config();
        List<String> path = new ArrayList<>();
        List<String> remaining = new ArrayList<>();
        boolean helpTreeSeen = false;

        for (int i = 0; i < args.length; ) {
            String arg = args[i];
            if ("--help-tree".equals(arg)) {
                config.helpTree = true;
                helpTreeSeen = true;
                i++;
            } else if (("-L".equals(arg) || "--tree-depth".equals(arg)) && i + 1 < args.length) {
                config.depthLimit = Integer.parseInt(args[++i]);
                i++;
            } else if (("-I".equals(arg) || "--tree-ignore".equals(arg)) && i + 1 < args.length) {
                config.ignoreNames.add(args[++i]);
                i++;
            } else if ("-a".equals(arg) || "--tree-all".equals(arg)) {
                config.showAll = true;
                i++;
            } else if ("--tree-output".equals(arg) && i + 1 < args.length) {
                config.outputFormat = args[++i];
                i++;
            } else if ("--tree-style".equals(arg) && i + 1 < args.length) {
                config.style = args[++i];
                i++;
            } else if ("--tree-color".equals(arg) && i + 1 < args.length) {
                config.color = args[++i];
                i++;
            } else if (!arg.startsWith("-") && !helpTreeSeen) {
                path.add(arg);
                i++;
            } else {
                remaining.add(arg);
                i++;
            }
        }

        if (config.helpTree) {
            config.path = path;
            config.remainingArgs = remaining;
        } else {
            List<String> all = new ArrayList<>(path);
            all.addAll(remaining);
            config.remainingArgs = all;
        }
        return config;
    }

    public static TreeCommand fromPicocli(CommandLine cmd) {
        CommandSpec spec = cmd.getCommandSpec();
        String name = spec.name();
        String description = "";
        if (spec.usageMessage().description() != null && spec.usageMessage().description().length > 0) {
            description = spec.usageMessage().description()[0];
        } else if (spec.usageMessage().header() != null && spec.usageMessage().header().length > 0) {
            description = spec.usageMessage().header()[0];
        }

        List<TreeOption> options = new ArrayList<>();
        for (OptionSpec opt : spec.options()) {
            String longName = null;
            String shortName = null;
            for (String n : opt.names()) {
                if (n.startsWith("--")) longName = n;
                else if (n.startsWith("-")) shortName = n;
            }
            if (longName == null) longName = shortName;

            if ("--help".equals(longName) || "-h".equals(shortName) ||
                "--version".equals(longName) || "-V".equals(shortName)) {
                continue;
            }

            String desc = "";
            if (opt.description() != null && opt.description().length > 0) {
                desc = opt.description()[0];
            }

            String optName = longName != null ? longName.replaceFirst("^--", "") : (shortName != null ? shortName.replaceFirst("^-", "") : "option");
            boolean takesValue = opt.arity().max > 0;

            options.add(new TreeOption(optName, shortName, longName, desc, opt.required(), takesValue, opt.hidden()));
        }

        List<TreeArgument> arguments = new ArrayList<>();
        for (PositionalParamSpec pos : spec.positionalParameters()) {
            String desc = "";
            if (pos.description() != null && pos.description().length > 0) {
                desc = pos.description()[0];
            }
            String argName = pos.paramLabel() != null ? pos.paramLabel() : "ARG";
            arguments.add(new TreeArgument(argName, desc, pos.arity().min > 0, pos.hidden()));
        }

        List<TreeCommand> subcommands = new ArrayList<>();
        for (java.util.Map.Entry<String, CommandLine> entry : cmd.getSubcommands().entrySet()) {
            if ("help".equals(entry.getKey())) continue;
            subcommands.add(fromPicocli(entry.getValue()));
        }

        return new TreeCommand(name, description, spec.usageMessage().hidden(), options, arguments, subcommands);
    }

    public static TreeCommand resolvePath(TreeCommand root, List<String> path) {
        TreeCommand current = root;
        for (String name : path) {
            TreeCommand next = null;
            for (TreeCommand child : current.subcommands) {
                if (child.name.equals(name)) {
                    next = child;
                    break;
                }
            }
            if (next == null) break;
            current = next;
        }
        return current;
    }

    public static String render(TreeCommand root, Config config) {
        Theme theme = resolveTheme(config);
        if ("json".equals(config.outputFormat)) {
            return toJson(root, config, 0);
        }
        StringBuilder sb = new StringBuilder();
        sb.append(theme.command.wrap(root.name)).append("\n");

        for (TreeOption opt : root.options) {
            if (opt.hidden && !config.showAll) continue;
            String meta;
            if (opt.shortName != null && opt.longName != null) {
                meta = opt.shortName + ", " + opt.longName;
            } else if (opt.longName != null) {
                meta = opt.longName;
            } else if (opt.shortName != null) {
                meta = opt.shortName;
            } else {
                meta = opt.name;
            }
            sb.append("  ").append(theme.options.wrap(meta))
              .append(" … ").append(theme.description.wrap(opt.description)).append("\n");
        }

        List<TreeCommand> children = filterChildren(root.subcommands, config);
        if (!children.isEmpty()) {
            sb.append("\n");
            for (int i = 0; i < children.size(); i++) {
                renderNode(children.get(i), "", i == children.size() - 1, config, theme, sb, 1);
            }
        }
        return sb.toString();
    }

    private static Theme resolveTheme(Config config) {
        boolean useColor = false;
        if ("always".equals(config.color)) {
            useColor = true;
        } else if ("auto".equals(config.color)) {
            useColor = System.console() != null;
        }
        if ("plain".equals(config.style)) {
            useColor = false;
        }
        return useColor ? Theme.defaultTheme() : Theme.plainTheme();
    }

    private static List<TreeCommand> filterChildren(List<TreeCommand> children, Config config) {
        List<TreeCommand> result = new ArrayList<>();
        for (TreeCommand child : children) {
            if (child.hidden && !config.showAll) continue;
            if (config.ignoreNames.contains(child.name)) continue;
            result.add(child);
        }
        return result;
    }

    private static void renderNode(TreeCommand node, String prefix, boolean isLast, Config config, Theme theme, StringBuilder sb, int depth) {
        if (config.depthLimit != null && depth > config.depthLimit) return;

        String branch = prefix + (isLast ? "\u2514\u2500\u2500 " : "\u251c\u2500\u2500 ");
        String name = node.name;
        String suffix = buildSuffix(node, config, depth);
        String desc = node.description != null ? node.description : "";

        int startCol = branch.length() + name.length() + suffix.length();
        int dotCount = Math.max(4, 32 - startCol);
        if (startCol >= 32) dotCount = 4;
        String dots = ".".repeat(dotCount);

        String line = branch
            + theme.command.wrap(name)
            + theme.options.wrap(suffix)
            + " " + dots + " "
            + theme.description.wrap(desc);

        sb.append(line).append("\n");

        if (config.depthLimit == null || depth < config.depthLimit) {
            List<TreeCommand> children = filterChildren(node.subcommands, config);
            for (int i = 0; i < children.size(); i++) {
                String childPrefix = prefix + (isLast ? "    " : "\u2502   ");
                renderNode(children.get(i), childPrefix, i == children.size() - 1, config, theme, sb, depth + 1);
            }
        }
    }

    private static String buildSuffix(TreeCommand node, Config config, int depth) {
        List<String> parts = new ArrayList<>();

        boolean hasFlags = node.options.stream().anyMatch(o -> {
            if (o.hidden && !config.showAll) return false;
            if (depth > 0 && isDiscoveryOption(o)) return false;
            return true;
        });
        if (hasFlags) parts.add("[flags]");

        for (TreeArgument arg : node.arguments) {
            if (arg.hidden && !config.showAll) continue;
            if (arg.required) parts.add("<" + arg.name + ">");
            else parts.add("[" + arg.name + "]");
        }

        if (parts.isEmpty()) return "";
        return " " + String.join(" ", parts);
    }

    private static boolean isDiscoveryOption(TreeOption o) {
        return "--help-tree".equals(o.longName) ||
               "--tree-depth".equals(o.longName) ||
               "--tree-ignore".equals(o.longName) ||
               "--tree-all".equals(o.longName) ||
               "--tree-output".equals(o.longName) ||
               "--tree-style".equals(o.longName) ||
               "--tree-color".equals(o.longName);
    }

    public static String toJson(TreeCommand node, Config config, int depth) {
        StringBuilder sb = new StringBuilder();
        sb.append("{");
        sb.append("\"type\":\"command\"");
        sb.append(",\"name\":\"").append(escapeJson(node.name)).append("\"");
        if (node.description != null && !node.description.isEmpty()) {
            sb.append(",\"description\":\"").append(escapeJson(node.description)).append("\"");
        }

        List<TreeOption> visibleOpts = new ArrayList<>();
        for (TreeOption o : node.options) {
            if (o.hidden && !config.showAll) continue;
            if (depth > 0 && isDiscoveryOption(o)) continue;
            visibleOpts.add(o);
        }

        if (!visibleOpts.isEmpty()) {
            sb.append(",\"options\":[");
            for (int i = 0; i < visibleOpts.size(); i++) {
                TreeOption o = visibleOpts.get(i);
                if (i > 0) sb.append(",");
                sb.append("{");
                sb.append("\"type\":\"option\"");
                sb.append(",\"name\":\"").append(escapeJson(o.name)).append("\"");
                if (o.shortName != null) sb.append(",\"short\":\"").append(escapeJson(o.shortName)).append("\"");
                if (o.longName != null) sb.append(",\"long\":\"").append(escapeJson(o.longName)).append("\"");
                if (o.description != null && !o.description.isEmpty()) {
                    sb.append(",\"description\":\"").append(escapeJson(o.description)).append("\"");
                }
                sb.append(",\"required\":").append(o.required);
                sb.append(",\"takes_value\":").append(o.takesValue);
                sb.append("}");
            }
            sb.append("]");
        }

        List<TreeArgument> visibleArgs = new ArrayList<>();
        for (TreeArgument a : node.arguments) {
            if (!a.hidden || config.showAll) visibleArgs.add(a);
        }
        if (!visibleArgs.isEmpty()) {
            sb.append(",\"arguments\":[");
            for (int i = 0; i < visibleArgs.size(); i++) {
                TreeArgument a = visibleArgs.get(i);
                if (i > 0) sb.append(",");
                sb.append("{");
                sb.append("\"type\":\"argument\"");
                sb.append(",\"name\":\"").append(escapeJson(a.name)).append("\"");
                if (a.description != null && !a.description.isEmpty()) {
                    sb.append(",\"description\":\"").append(escapeJson(a.description)).append("\"");
                }
                sb.append(",\"required\":").append(a.required);
                sb.append("}");
            }
            sb.append("]");
        }

        if (config.depthLimit == null || depth < config.depthLimit) {
            List<TreeCommand> children = filterChildren(node.subcommands, config);
            if (!children.isEmpty()) {
                sb.append(",\"subcommands\":[");
                for (int i = 0; i < children.size(); i++) {
                    if (i > 0) sb.append(",");
                    sb.append(toJson(children.get(i), config, depth + 1));
                }
                sb.append("]");
            }
        }

        sb.append("}");
        return sb.toString();
    }

    private static String escapeJson(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }
}

File: kotlin/src/main/kotlin/helptree/HelpTree.kt
package helptree

import java.io.Console

data class TextTokenTheme(
    val emphasis: String = "normal",
    val colorHex: String? = null
)

data class Theme(
    val command: TextTokenTheme = TextTokenTheme("bold", "#7ee7e6"),
    val options: TextTokenTheme = TextTokenTheme(),
    val description: TextTokenTheme = TextTokenTheme("italic", "#90a2af")
)

data class TreeOption(
    val name: String = "",
    val short: String = "",
    val long: String = "",
    val description: String = "",
    val required: Boolean = false,
    val takesValue: Boolean = false,
    val defaultVal: String = "",
    val hidden: Boolean = false
)

data class TreeArgument(
    val name: String = "",
    val description: String = "",
    val required: Boolean = false,
    val hidden: Boolean = false
)

data class TreeCommand(
    val name: String = "",
    val description: String = "",
    val options: List<TreeOption> = emptyList(),
    val arguments: List<TreeArgument> = emptyList(),
    val subcommands: List<TreeCommand> = emptyList(),
    val hidden: Boolean = false
)

data class Config(
    var helpTree: Boolean = false,
    var depthLimit: Int? = null,
    var ignoreNames: List<String> = emptyList(),
    var showAll: Boolean = false,
    var outputFormat: String = "text",
    var style: String = "rich",
    var color: String = "auto",
    var path: List<String> = emptyList(),
    var remainingArgs: List<String> = emptyList()
)

fun discoveryOptions(): List<TreeOption> = listOf(
    TreeOption("help-tree", long = "--help-tree", description = "Print a recursive command map derived from framework metadata"),
    TreeOption("tree-depth", short = "-L", long = "--tree-depth", description = "Limit --help-tree recursion depth (Unix tree -L style)", takesValue = true),
    TreeOption("tree-ignore", short = "-I", long = "--tree-ignore", description = "Exclude subtrees/commands from --help-tree output (repeatable)", takesValue = true),
    TreeOption("tree-all", short = "-a", long = "--tree-all", description = "Include hidden subcommands in --help-tree output"),
    TreeOption("tree-output", long = "--tree-output", description = "Output format (text or json)", takesValue = true),
    TreeOption("tree-style", long = "--tree-style", description = "Tree text styling mode (rich or plain)", takesValue = true),
    TreeOption("tree-color", long = "--tree-color", description = "Tree color mode (auto, always, never)", takesValue = true)
)

fun extractConfig(args: Array<String>): Config {
    val config = Config()
    val path = mutableListOf<String>()
    val remaining = mutableListOf<String>()
    var helpTreeSeen = false

    var i = 0
    while (i < args.size) {
        val arg = args[i]
        when {
            arg == "--help-tree" -> {
                config.helpTree = true
                helpTreeSeen = true
                i++
            }
            (arg == "-L" || arg == "--tree-depth") && i + 1 < args.size -> {
                config.depthLimit = args[++i].toIntOrNull()
                i++
            }
            (arg == "-I" || arg == "--tree-ignore") && i + 1 < args.size -> {
                config.ignoreNames = config.ignoreNames + args[++i]
                i++
            }
            arg == "-a" || arg == "--tree-all" -> {
                config.showAll = true
                i++
            }
            arg == "--tree-output" && i + 1 < args.size -> {
                config.outputFormat = args[++i]
                i++
            }
            arg == "--tree-style" && i + 1 < args.size -> {
                config.style = args[++i]
                i++
            }
            arg == "--tree-color" && i + 1 < args.size -> {
                config.color = args[++i]
                i++
            }
            !arg.startsWith("-") && !helpTreeSeen -> {
                path.add(arg)
                i++
            }
            else -> {
                remaining.add(arg)
                i++
            }
        }
    }

    if (config.helpTree) {
        config.path = path
        config.remainingArgs = remaining
    } else {
        config.remainingArgs = path + remaining
    }
    return config
}

fun shouldUseColor(config: Config): Boolean {
    return when (config.color) {
        "always" -> true
        "never" -> false
        else -> System.console() != null
    }
}

fun useRichStyle(config: Config): Boolean {
    return config.style != "plain"
}

fun parseHexRgb(hex: String): Triple<Int, Int, Int>? {
    val h = hex.removePrefix("#")
    if (h.length != 6) return null
    return try {
        Triple(
            h.substring(0, 2).toInt(16),
            h.substring(2, 4).toInt(16),
            h.substring(4, 6).toInt(16)
        )
    } catch (_: NumberFormatException) {
        null
    }
}

fun styleText(text: String, token: TextTokenTheme, config: Config): String {
    if (!useRichStyle(config) || (token.emphasis == "normal" && token.colorHex == null)) {
        return text
    }
    val codes = mutableListOf<String>()
    when (token.emphasis) {
        "bold" -> codes.add("1")
        "italic" -> codes.add("3")
        "bold_italic" -> { codes.add("1"); codes.add("3") }
    }
    if (shouldUseColor(config) && token.colorHex != null) {
        val rgb = parseHexRgb(token.colorHex)
        if (rgb != null) {
            codes.add("38;2;${rgb.first};${rgb.second};${rgb.third}")
        }
    }
    if (codes.isEmpty()) return text
    return "\u001B[${codes.joinToString(";")}m$text\u001B[0m"
}

fun shouldSkipOption(opt: TreeOption, treeAll: Boolean): Boolean {
    if (treeAll) return false
    if (opt.hidden) return true
    if (opt.name == "help" || opt.name == "version") return true
    return false
}

fun shouldSkipArgument(arg: TreeArgument, treeAll: Boolean): Boolean {
    if (treeAll) return false
    if (arg.hidden) return true
    return false
}

fun shouldSkipCommand(cmd: TreeCommand, config: Config): Boolean {
    if (cmd.name == "help") return true
    if (config.ignoreNames.contains(cmd.name)) return true
    if (!config.showAll && cmd.hidden) return true
    return false
}

fun isDiscoveryOption(opt: TreeOption): Boolean {
    return opt.long == "--help-tree" ||
            opt.long == "--tree-depth" ||
            opt.long == "--tree-ignore" ||
            opt.long == "--tree-all" ||
            opt.long == "--tree-output" ||
            opt.long == "--tree-style" ||
            opt.long == "--tree-color"
}

fun commandSignature(cmd: TreeCommand, treeAll: Boolean): Pair<String, String> {
    val suffix = StringBuilder()
    cmd.arguments.forEach { arg ->
        if (!shouldSkipArgument(arg, treeAll)) {
            suffix.append(if (arg.required) " <${arg.name}>" else " [${arg.name}]")
        }
    }
    val hasFlags = cmd.options.any { !shouldSkipOption(it, treeAll) }
    if (hasFlags) suffix.append(" [flags]")
    return cmd.name to suffix.toString()
}

fun renderTextLines(cmd: TreeCommand, prefix: String, depth: Int, config: Config, theme: Theme): List<String> {
    val items = cmd.subcommands.filter { !shouldSkipCommand(it, config) }
    if (items.isEmpty()) return emptyList()

    val atLimit = config.depthLimit?.let { depth >= it } ?: false
    val lines = mutableListOf<String>()

    items.forEachIndexed { i, sub ->
        val isLast = i == items.size - 1
        val branch = if (isLast) "└── " else "├── "
        val (name, suffix) = commandSignature(sub, config.showAll)
        val signature = name + suffix
        val about = sub.description
        val sigStyled = styleText(name, theme.command, config) + styleText(suffix, theme.options, config)

        val line = if (about.isNotEmpty()) {
            val dotsLen = maxOf(4, 28 - signature.length)
            val dots = ".".repeat(dotsLen)
            "$prefix$branch$sigStyled $dots ${styleText(about, theme.description, config)}"
        } else {
            "$prefix$branch$sigStyled"
        }
        lines.add(line)

        if (!atLimit) {
            val extension = if (isLast) "    " else "│   "
            lines.addAll(renderTextLines(sub, prefix + extension, depth + 1, config, theme))
        }
    }
    return lines
}

fun render(cmd: TreeCommand, config: Config): String {
    if (config.outputFormat == "json") {
        return toJson(cmd, config, 0)
    }

    val theme = Theme()
    val lines = mutableListOf<String>()
    lines.add(styleText(cmd.name, theme.command, config))

    cmd.options.forEach { opt ->
        if (!shouldSkipOption(opt, config.showAll)) {
            val meta = when {
                opt.short.isNotEmpty() && opt.long.isNotEmpty() -> "${opt.short}, ${opt.long}"
                opt.long.isNotEmpty() -> opt.long
                opt.short.isNotEmpty() -> opt.short
                else -> opt.name
            }
            lines.add("  ${styleText(meta, theme.options, config)} … ${styleText(opt.description, theme.description, config)}")
        }
    }

    val treeLines = renderTextLines(cmd, "", 0, config, theme)
    if (treeLines.isNotEmpty()) {
        lines.add("")
        lines.addAll(treeLines)
    }

    lines.add("")
    lines.add("Use `${cmd.name} <COMMAND> --help` for full details on arguments and flags.")
    return lines.joinToString("\n")
}

fun toJson(cmd: TreeCommand, config: Config, depth: Int): String {
    val sb = StringBuilder()
    sb.append("{")
    sb.append("\"type\":\"command\"")
    sb.append(",\"name\":\"${escapeJson(cmd.name)}\"")
    if (cmd.description.isNotEmpty()) {
        sb.append(",\"description\":\"${escapeJson(cmd.description)}\"")
    }

    val visibleOpts = cmd.options.filter { !shouldSkipOption(it, config.showAll) }
    if (visibleOpts.isNotEmpty()) {
        sb.append(",\"options\":[")
        visibleOpts.forEachIndexed { i, opt ->
            if (i > 0) sb.append(",")
            sb.append("{")
            sb.append("\"type\":\"option\"")
            sb.append(",\"name\":\"${escapeJson(opt.name)}\"")
            if (opt.description.isNotEmpty()) sb.append(",\"description\":\"${escapeJson(opt.description)}\"")
            if (opt.short.isNotEmpty()) sb.append(",\"short\":\"${escapeJson(opt.short)}\"")
            if (opt.long.isNotEmpty()) sb.append(",\"long\":\"${escapeJson(opt.long)}\"")
            if (opt.defaultVal.isNotEmpty()) sb.append(",\"default\":\"${escapeJson(opt.defaultVal)}\"")
            sb.append(",\"required\":${opt.required}")
            sb.append(",\"takes_value\":${opt.takesValue}")
            sb.append("}")
        }
        sb.append("]")
    }

    val visibleArgs = cmd.arguments.filter { !shouldSkipArgument(it, config.showAll) }
    if (visibleArgs.isNotEmpty()) {
        sb.append(",\"arguments\":[")
        visibleArgs.forEachIndexed { i, arg ->
            if (i > 0) sb.append(",")
            sb.append("{")
            sb.append("\"type\":\"argument\"")
            sb.append(",\"name\":\"${escapeJson(arg.name)}\"")
            if (arg.description.isNotEmpty()) sb.append(",\"description\":\"${escapeJson(arg.description)}\"")
            sb.append(",\"required\":${arg.required}")
            sb.append("}")
        }
        sb.append("]")
    }

    val canRecurse = config.depthLimit == null || depth < config.depthLimit!!
    if (canRecurse) {
        val children = cmd.subcommands.filter { !shouldSkipCommand(it, config) }
        if (children.isNotEmpty()) {
            sb.append(",\"subcommands\":[")
            children.forEachIndexed { i, child ->
                if (i > 0) sb.append(",")
                sb.append(toJson(child, config, depth + 1))
            }
            sb.append("]")
        }
    }

    sb.append("}")
    return sb.toString()
}

fun escapeJson(s: String): String {
    return s.replace("\\", "\\\\")
        .replace("\"", "\\\"")
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
}

fun resolvePath(root: TreeCommand, path: List<String>): TreeCommand {
    var current = root
    for (name in path) {
        val next = current.subcommands.find { it.name == name }
        if (next == null) break
        current = next
    }
    return current
}
